pkg/sql/copy_file_upload.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/blobs"
+	"github.com/cockroachdb/cockroach/pkg/internal/client"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgwirebase"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/lib/pq"
+)
+
+const (
+	fileUploadTable = "crdb_internal.file_upload"
+	copyOptionDest  = "destination"
+)
+
+var copyFileOptionExpectValues = map[string]KVStringOptValidate{
+	copyOptionDest: KVStringOptRequireValue,
+}
+
+type copyError struct {
+	err error
+}
+
+var _ copyMachineInterface = &fileUploadMachine{}
+
+type fileUploadMachine struct {
+	c           *copyMachine
+	writeToFile *io.PipeWriter
+	writeError  chan copyError
+}
+
+func newFileUploadMachine(
+	conn pgwirebase.Conn,
+	n *tree.CopyFrom,
+	execCfg *ExecutorConfig,
+	resetPlanner func(p *planner, txn *client.Txn, txnTS time.Time, stmtTS time.Time),
+) (f *fileUploadMachine, retErr error) {
+	c := &copyMachine{
+		conn:    conn,
+		table:   &n.Table,
```

Comment:
For the file uploader, this and columns can be unset because they aren't needed, so we shouldn't set them. If they are empty and required, it might cause a loud error, which is good. (There's some refactoring work that could be done here to remove the dependency on copyMachine.tables/columns from insertRows where it's currently used, but only do that if you're feeling ambitious.)

Reply:
Okay, will add a TODO to remove this dependency from insertRows.